src/fillalgebra.jl

```diff
@@ -160,38 +160,33 @@ function *(a::Transpose{T, <:AbstractVector{T}}, b::ZerosVector{T}) where T<:Rea
 end
 *(a::Transpose{T, <:AbstractMatrix{T}}, b::ZerosVector{T}) where T<:Real = mult_zeros(a, b)
 
-# treat zero separately to support ∞-vectors
-function _zero_dot(a, b)
-    axes(a) == axes(b) || throw(DimensionMismatch("dot product arguments have lengths $(length(a)) and $(length(b))"))
-    zero(promote_type(eltype(a),eltype(b)))
-end
-
-_fill_dot(a::Zeros, b::Zeros) = _zero_dot(a, b)
-_fill_dot(a::Zeros, b) = _zero_dot(a, b)
-_fill_dot(a, b::Zeros) = _zero_dot(a, b)
-_fill_dot(a::Zeros, b::AbstractFill) = _zero_dot(a, b)
-_fill_dot(a::AbstractFill, b::Zeros) = _zero_dot(a, b)
-
-function _fill_dot(a::AbstractFill, b::AbstractFill)
-    axes(a) == axes(b) || throw(DimensionMismatch("dot product arguments have lengths $(length(a)) and $(length(b))"))
-    getindex_value(a)getindex_value(b)*length(b)
-end
-
 # support types with fast sum
-function _fill_dot(a::AbstractFill, b)
+
+function _fill_dot(a::AbstractFillVector{T}, b::AbstractVector{V}) where {T,V}
     axes(a) == axes(b) || throw(DimensionMismatch("dot product arguments have lengths $(length(a)) and $(length(b))"))
-    getindex_value(a)sum(b)
+    S = promote_op(+, T, V)
+    # treat zero separately to support ∞-vectors
+    if iszero(b) || iszero(a)
+        zero(S)
+    else
+        dot(getindex_value(a), S(sum(b)))
+    end
 end
 
-function _fill_dot(a, b::AbstractFill)
+function _fill_dot_rev(a::AbstractVector{T}, b::AbstractFillVector{V}) where {T,V}
     axes(a) == axes(b) || throw(DimensionMismatch("dot product arguments have lengths $(length(a)) and $(length(b))"))
-    sum(a)getindex_value(b)
+    S = promote_op(+, T, V)
+    # treat zero separately to support ∞-vectors
+    if iszero(b) || iszero(a)
+        zero(S)
+    else
+        dot(S(sum(a)), getindex_value(b))
```

Comment:
```julia
       convert(S, dot(sum(a), getindex_value(b)))
```
Just to make sure. Also, using `convert` will avoid allocations.

Reply:
I'm thinking of making changes here. Have you read https://github.com/JuliaArrays/FillArrays.jl/pull/229#issuecomment-1488566501 ?